### DISCUSSION TITLE
Physicist noob question about arbitrary PCB waveguides

### DISCUSSION THREAD

**Discussion by settwi** [thliebig/openEMS-Project#205] [Q&A Usage]
Hi there, I am working to design a photodetector circuit which operates at relatively high frequencies (~1 GHz).

I want to send my amplified photon pulses through some waveguides. It would be great if those could be coplanar waveguides with shielding vias.

**Can I define an arbitrary geometry waveguide with different dielectrics, etc.?** I was looking at [atlc](https://atlc.sourceforge.net/) but it gets quite slow for large images. I'm attaching an example image of what I was imagining to make. Green is the ground conductor, red is the live one, white is vacuum, and gray is the PCB dielectric.
![example](https://github.com/thliebig/openEMS-Project/assets/13972006/16065c1b-9b6b-485c-b18f-1c1a64d2d62c)

Thanks!

**Comment by VolkerMuehlhaus** [comment]
> Can I define an arbitrary geometry waveguide with different dielectrics, etc.?

Yes, this is possible.

But I don't think that solving the 3D volume mesh model with openEMS is faster than solving the 2D cross section using atlc.

**Comment by settwi** [comment]
it seems like atlc is actually quite sensitive to the size precision on the different portions. the example i attached goes from like 58 to ~50 ohms for smaller to larger images. and it is extremely slow to process the larger images...

**Reply by VolkerMuehlhaus** [reply]
Yes, these numericals methods are sensitive to mesh resolution, and I guess that atlc will create the mesh based on bitmap resolution. 

In real world, the CPW line impedance changes quite a bit when you vary the metal thickness, so that could explain some of the effect that you see.

I don't know if you want to use openEMS with the Matlab oder Python user interface? 
For Matlab there is a CPW example included in matlab\examples\transmission_lines
For Python there is no CPW example, so you would start from the microstrip filter example python\Tutorials\MSL_NotchFilter.py to get familiar with the basics, and then build CPW by grabbing information from the source code comments, and from the other threads found here.

**Comment by biergaizi** [comment]
For 2D transmission line problems, another option to try is TNT-MMTL. In my opinion it's probably one of the best free 2D transmission line field solvers. It was developed during the 1980s by Mayo Clinic’s Special Purpose Processor Development Group. Development stopped and GPL'd in 2004 - it's ancient but it still works. It's a very basic quasi-static solver and cannot solve waveguide modes (TEM only), but usually the results are within a few percent within proprietary commercial tools. There are a few tutorials on YouTube, just search TNT-MMTL, such as [0]. It can be difficult to build it on a modern Unix/Linux system, so if you have zero patience, it's easier to use the Win32 installer. [1] I was actually working on a WebAssembly port so you can run the solver directly on a Web browser, but didn't have time to finish it - as now I'm working on optimizing openEMS instead [2].

For openEMS modeling, here's an example of a previous attempt, proposed by me and worked out by @VolkerMuehlhaus. [3]

[0] https://www.youtube.com/watch?v=kzst_wpsXWY

[1] But it's still possible on run Unix. I have a personal fork with all the necessary patches applied: https://notabug.org/niconiconi/tnt-mmtl, use the `main` branch, `./configure` and `make`. Also make sure to install packages Tcl, Tk, Incr Tcl, IWidgets, and BWidget via your package manager, name may vary, just search for similar names.

[2] https://niconiconi.neocities.org/tech-notes/fortran-in-webassembly-and-field-solver/

[3] https://github.com/thliebig/openEMS-Project/discussions/38

**Comment by settwi** [comment]
Awesome, thank you for the resources! I will look into them
